File: InterviewPrep/IsRotation.java
import java.util.*;
public class IsRotation {
    public static void main(String[] args) {
        // NOTE: The following input values will be used for testing your solution.
        int[] array1 = {1, 2, 3, 4, 5, 6, 7};
        int[] array2a = {4, 5, 6, 7, 8, 1, 2, 3};
        System.out.println(isRotation(array1, array2a)); // should return false.
        int[] array2b = {4, 5, 6, 7, 1, 2, 3};
        System.out.println(isRotation(array1, array2b)); //should return true.
        int[] array2c = {4, 5, 6, 9, 1, 2, 3};
        // isRotation(array1, array2c) should return false.
        int[] array2d = {4, 6, 5, 7, 1, 2, 3};
        // isRotation(array1, array2d) should return false.
        int[] array2e = {4, 5, 6, 7, 0, 2, 3};
        // isRotation(array1, array2e) should return false.
        int[] array2f = {1, 2, 3, 4, 5, 6, 7};
        // isRotation(array1, array2f) should return true.
    }

    // Implement your solution below.
    public static Boolean isRotation(int[] array1, int[] array2) {
        if (array1.length != array2.length) {
            return false;
        }
        if (array1.length == 0 || array1.length == 1) {
            return true;
        }
        ArrayList<Integer> arr2Indices = new ArrayList<Integer>();
        for (int i = 0; i < array2.length; i++) {
            if (array2[i] == array1[0]) {
                arr2Indices.add(i);
            }
        }
        int count = 0;
        for (int index: arr2Indices) {
            count = 0;
            int diff = index; // index - 0, because the index we're looking at in arr1 is 0.
            for (int j = 0; j < array1.length; j++) {
                if (array1[j] != array2[(j + diff) % array1.length]) {
                    break;
                }
                count++;
            }
            if (count == array1.length) {
                return true;
            }
        }
        return false;
    }
}

File: InterviewPrep/isRotation.py
# Implement your function below.
# Not a very good implementation, but correct.

def is_rotation(l1, l2):
    # Ex: [1,2,3,4,5] and [3,4,5,1,2]: True
    #     [1,3,4,9] and [4,9,1,8]: False
    if len(l1) != len(l2):
        return False
    # Equal length. 
    if len(l1) == 0 or len(l1) == 1:
        return True
    # Choose the first elem in list1, go through list2 to find the first occurrence and save the index in a var.
    # Check from that index onwards by (incrementing mod n) if elements are the same. If not, increment stored index till n, keep trying.
    elem = l1[0]
    index_l1 = 0
    l2_indices = []
    for i in range(len(l2)):
        if l2[i] == elem:
            l2_indices.append(i)
    count = 0
    for index_l2 in l2_indices:
        count = 0
        diff = index_l2 - index_l1
        for i in range(len(l1)):
            if l1[i] != l2[(i + diff) % len(l1)]:
                break
            count += 1
        if count == len(l1):
            return True
            
    return False

# NOTE: The following input values will be used for testing your solution.
list1 = [1, 2, 3, 4, 5, 6, 7]
list2a = [4, 5, 6, 7, 8, 1, 2, 3]
print is_rotation(list1, list2a) #should return False.
list2b = [4, 5, 6, 7, 1, 2, 3]
print is_rotation(list1, list2b) #should return True.
list2c = [4, 5, 6, 9, 1, 2, 3]
print is_rotation(list1, list2c) #should return False.
list2d = [4, 6, 5, 7, 1, 2, 3]
print is_rotation(list1, list2d) #should return False.
list2e = [4, 5, 6, 7, 0, 2, 3]
print is_rotation(list1, list2e) #should return False.
list2f = [1, 2, 3, 4, 5, 6, 7]
print is_rotation(list1, list2f) #should return True.

